homeassistant/components/essent/sensor.py

```diff
@@ -63,10 +69,14 @@ def retrieve_meter_data(self, meter):
     def update(self):
         """Retrieve the latest meter data from Essent."""
         essent = PyEssent(self._username, self._password)
-        self._meters = essent.get_EANs()
-        for meter in self._meters:
-            self._meter_data[meter] = essent.read_meter(
-                meter, only_last_meter_reading=True)
+        EANs = essent.get_EANs()
+        for possible_meter in EANs:
+            meter_data = essent.read_meter(
+                possible_meter, only_last_meter_reading=True)
+            if meter_data:
+                if possible_meter not in self._meters:
+                    self._meters.append(possible_meter)
```

Comment:
Would it make sense to implement the `retrieve_meters` method as `self._meter_data.keys()` instead of maintaining the meter IDs twice ?

Reply:
You're completely right, fixed.